adding announcement for release v0.3.0 (#465)
kudobuilder/kudo: site/content/posts/announcing-kudo-0.3.0.md

---
title: Announcing KUDO 0.3.0
date: 2019-06-28
---

> We are proud to announce the release of KUDO v0.3.0! This release focuses on core infrastructure inside of KUDO as the format for developing operators and running operators stabilizes.

Reviewer: This is the same sentence we used in the v0.2.0 announcement. For a second I felt lost. Let's give a one-sentence summary of what's new instead.

## What is KUDO?

[Kubernetes Universal Declarative Operator (KUDO)](https://github.com/kudobuilder/kudo) provides a declarative approach to building production-grade Kubernetes Operators covering the entire application lifecycle. An operator is a way to package and manage a Kubernetes application using Kubernetes APIs. Building an Operator usually involves implementing a [custom resource and controller](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/), which typically requires thousands of lines of code and a deep understanding of Kubernetes. KUDO instead provides a universal controller that can be configured via a declarative spec (YAML files) to operate any workload.

## Release Highlights

### Switch from Framework to Operator

> Framework was the term the project previously used to represent a service that KUDO "the operator" would be instrumented to be an operator for. The intention was to bring clarity by introducing a new term. It was thought that the new term would allow for more accuracy in technical communication by reduce the overloading of the term operator. For some it caused more confusion and almost always resulted in the need for explanation. We have abandon the term completely resulting in a **breaking** change to the CRDs.

Reviewer: by reduce => by reducing
have abandon => have abandoned

> All previous concepts and required objects using the word Framework have been replaced with Operator, resulting Operator and OperatorVersion in the type definition.

Reviewer: resulting => resulting in

> This will result in some overloading of the term operator, but aligns better with the eco-system. The running KUDO controller is an operator that manages operators built with KUDO's packaging format. In the future, this will be expanded with support for Helm charts (v0.4.0), other packaging formats, existing operators, and beyond. KUDO's mission is to enable easy, declarative day two operations and API management for complex applications, no matter how they were originally built.

Reviewer: day two operations => Day 2 operations (that's the most common spelling)

### Package Format

> [KEP-10](https://github.com/kudobuilder/kudo/blob/v0.3.0/keps/0010-package-manager.md) for the package manager is underway, with release providing 2 features; package repository and a package format.

Reviewer: with release => with this release

Reviewer: 2 features => two features

The general format allows for a number of files, but starts with an `operator.yaml`. A common layout might be:

```
operator.yaml
params.yaml
templates/
     <additional files>.yaml
```
The `operator.yaml` defines the operator and its plans. `params.yaml` defines the params to be used through all templated files. While the templated files are defined under the templates folder. The templated files are referenced in the `operator.yaml` file. The zookeeper operator provides a [good example](https://github.com/kudobuilder/operators/tree/a1f4cf/repository/zookeeper/0.1.0).

> These package files are group together in a folder or subfolder or can be tarred together in a `tgz` file providing a couple of ways to install an operator.

Reviewer: are group => are grouped

### Package Installation Options

> Introduced with this release are 2 ways to install an operator to KUDO. The first is designed for operator developers which is to install from the local filesystem. This can be either a folder with the operator package files or a tarball of those files.

Reviewer: 2 ways => two ways

> Introduced with this released are 2 ways to install an operator to KUDO. The first is designed for operator developers which is to install from the local filesystem. This can be either a folder with the operator package files or a tarball of those files.

Reviewer 2: release

Reviewer 2: ```suggestion
Introduced with this release are 2 ways to install an operator to KUDO. The first is designed for operator developers which is to install from the local filesystem. This can be either a folder with the operator package files or a tarball of those files.
```

> `kubectl kudo install ./kafka` or `kubectl kudo install kafka.tgz`

Reviewer: change the formatting so each command is on its own line. Easier to read that way.

This release also introduces a repository which allows for automatic discovery of operators to be installed.

> `kubectl kudo install kakfa`

Reviewer: change formatting so command is on its own line

**Note:** The order of operation when installing an operator is to check the local filesystem first, then to query the repository.

### Skip Instance

> kudoctl install now supports `-skip-instance` flag which skips installing an Instance when installing Operator and OperatorVersions. This is useful in a testing scenario - we install all of the Operators and OperatorVersions but don't want to install the Instances until later.

Reviewer: -skip-instance => --skip-instance (missing a dash)

### Test Harness

> The test harness outlined in [KEP-8](https://github.com/kudobuilder/kudo/blob/v0.3.0/keps/0008-operator-testing.md) has been implemented.  The harness described in [testing.md](https://github.com/kudobuilder/kudo/blob/v0.3.0/docs/testing.md) provides a mechanism for operator developers to create end to end integration testing for their frameworks.  Introduced with the test harness is the `kind: TestAssert` to assert operator state.  [Example tests](https://github.com/kudobuilder/operators/tree/v0.3.0/repository/zookeeper/tests/zookeeper-upgrade-test) are provided for the zookeeper operator.

Reviewer 2: Maybe change the link to https://kudo.dev/docs/testing

Author: that url is above for the word harness that could definitely be improved.

Reviewer 2: i included this in my suggested change, so it should be updated already

> The test harness outlined in [KEP-8](https://github.com/kudobuilder/kudo/blob/v0.3.0/keps/0008-operator-testing.md) has been implemented.  The harness described in [testing.md](https://github.com/kudobuilder/kudo/blob/v0.3.0/docs/testing.md) provides a mechanism for operator developers to create end to end integration testing for their frameworks.  Introduced with the test harness is the `kind: TestAssert` to assert operator state.  [Example tests](https://github.com/kudobuilder/operators/tree/v0.3.0/repository/zookeeper/tests/zookeeper-upgrade-test) are provided for the zookeeper operator.

Reviewer 2: I'd say let's drop this sentence about the TestAssert as it might be confusing.

Reviewer 2: Last thing, can we change this sentence:

```
The harness described in [testing.md](https://github.com/kudobuilder/kudo/blob/v0.3.0/docs/testing.md) provides a mechanism for operator developers to create end to end integration testing for their frameworks.
```

To:

```
The [harness](https://kudo.dev/docs/testing) provides a mechanism for operator developers to declaratively create end to end integration testing for their frameworks using only Kubernetes YAML.
```

> The test harness outlined in [KEP-8](https://github.com/kudobuilder/kudo/blob/v0.3.0/keps/0008-operator-testing.md) has been implemented.  The harness described in [testing.md](https://github.com/kudobuilder/kudo/blob/v0.3.0/docs/testing.md) provides a mechanism for operator developers to create end to end integration testing for their frameworks.  Introduced with the test harness is the `kind: TestAssert` to assert operator state.  [Example tests](https://github.com/kudobuilder/operators/tree/v0.3.0/repository/zookeeper/tests/zookeeper-upgrade-test) are provided for the zookeeper operator.

Reviewer 2: ```suggestion
The test harness outlined in [KEP-8](https://github.com/kudobuilder/kudo/blob/v0.3.0/keps/0008-operator-testing.md) has been implemented.  The [harness](https://kudo.dev/docs/testing) provides a mechanism for operator developers to declaratively create end to end integration testing for their operators using only Kubernetes YAML.  [Example tests](https://github.com/kudobuilder/operators/tree/v0.3.0/repository/zookeeper/tests/zookeeper-upgrade-test) are provided for the zookeeper operator.
```

### Package Installation Options

Introduced with this release are two ways to install an operator to KUDO. The first is designed for operator developers which is to install from the local filesystem. This can be either a folder with the operator package files or a tarball of those files.

To install an operator from a folder run:

`kubectl kudo install ./my-kafka-folder`

To install from a tarball run:

`kubectl kudo install kafka.tgz`

This release also introduces a repository which allows for automatic discovery of operators to be installed. To install from the repository run:

`kubectl kudo install kakfa`

**Note:** The order of operation when installing an operator is to check the local filesystem first, then to query the repository.

### Skip Instance

kudoctl install now supports `--skip-instance` flag which skips installing an Instance when installing Operator and OperatorVersions. This is useful in a testing scenario - we install all of the Operators and OperatorVersions but don't want to install the Instances until later.

### Test Harness

> The test harness outlined in [KEP-8](https://github.com/kudobuilder/kudo/blob/v0.3.0/keps/0008-operator-testing.md) has been implemented. The [test harness](https://kudo.dev/docs/testing) provides a mechanism for operator developers to declaratively create end-to-end integration testing for their frameworks using only Kubernetes YAML. [Example tests](https://github.com/kudobuilder/operators/tree/v0.3.0/repository/zookeeper/tests/zookeeper-upgrade-test) are provided for the zookeeper operator.

Reviewer 3: ```suggestion
The test harness outlined in [KEP-8](https://github.com/kudobuilder/kudo/blob/v0.3.0/keps/0008-operator-testing.md) has been implemented. The [test harness](https://kudo.dev/docs/testing) provides a mechanism for operator developers to declaratively create end-to-end integration tests for their frameworks using only Kubernetes YAML. [Example tests](https://github.com/kudobuilder/operators/tree/v0.3.0/repository/zookeeper/tests/zookeeper-upgrade-test) are provided for the zookeeper operator.
```

> The test harness outlined in [KEP-8](https://github.com/kudobuilder/kudo/blob/v0.3.0/keps/0008-operator-testing.md) has been implemented. The [test harness](https://kudo.dev/docs/testing) provides a mechanism for operator developers to declaratively create end-to-end integration testing for their frameworks using only Kubernetes YAML. [Example tests](https://github.com/kudobuilder/operators/tree/v0.3.0/repository/zookeeper/tests/zookeeper-upgrade-test) are provided for the zookeeper operator.

Reviewer 4: ```suggestion
The test harness outlined in [KEP-8](https://github.com/kudobuilder/kudo/blob/v0.3.0/keps/0008-operator-testing.md) has been implemented. The [test harness](https://kudo.dev/docs/testing) provides a mechanism for operator developers to declaratively create end-to-end integration tests for their operators using only Kubernetes YAML. [Example tests](https://github.com/kudobuilder/operators/tree/v0.3.0/repository/zookeeper/tests/zookeeper-upgrade-test) are provided for the zookeeper operator.
```
